src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10755,97 +10844,127 @@ private ExpressionSyntax ParseRegularStackAllocExpression()
 
         private AnonymousMethodExpressionSyntax ParseAnonymousMethodExpression()
         {
-            bool parentScopeIsInAsync = IsInAsync;
-            IsInAsync = false;
-            SyntaxToken asyncToken = null;
-            if (this.CurrentToken.ContextualKind == SyntaxKind.AsyncKeyword)
+            var parentScopeIsInAsync = this.IsInAsync;
+            var result = parseAnonymousMethodExpressionWorker();
+            this.IsInAsync = parentScopeIsInAsync;
+            return result;
+
+            AnonymousMethodExpressionSyntax parseAnonymousMethodExpressionWorker()
             {
-                asyncToken = this.EatContextualToken(SyntaxKind.AsyncKeyword);
-                asyncToken = CheckFeatureAvailability(asyncToken, MessageID.IDS_FeatureAsync);
-                IsInAsync = true;
-            }
+                var modifiers = ParseAnonymousFunctionModifiers();
+                if (modifiers.Any((int)SyntaxKind.AsyncKeyword))
+                {
+                    this.IsInAsync = true;
```

Comment:
>this.IsInAsync = true; [](start = 20, length = 22)

Don't we need to set `IsInAsync` unconditionally?
```
IsInAsync = modifiers.Any((int)SyntaxKind.AsyncKeyword);
```

Reply:
no.  'async' is only 'or-ed' in.  Once it's true at a higher level, it can't become false deeper down.  I actually made the mistake of doing it unconditionally, and that broke tests that explicit check for this.